#Csharp 

# Parallel LINQ, Обработка ошибок и отмена параллельных операций в C#

При возникновении ошибки в одном потоке выполнение всех потоков прекращается.
- `AggregateException` - перехватывает исключения из всех потоков, складывая в список  `InnerExceptions`.
- `WithCancellation()` - позволяет прерывать параллельный запрос, принимая `CancellationToken`, который прерывает операцию при вызове `Cancel()`.
- Исключение `OperationCanceledException` - возникает при прерывании параллельной операции (попадает в `AggregateException` в случае также других ошибок, что следует учитывать при обработке).

---

### Примеры кода:

1. **Обработка ошибок с использованием `AggregateException`** — демонстрирует агрегацию исключений при неудачном преобразовании:
    ```csharp
    object[] numbers = new object[] { 1, 2, 3, 4, 5, "6" };

    var squares = from n in numbers.AsParallel()
                     let x = (int)n
                     select Square(x);
    try
    {
        squares.ForAll(n => Console.WriteLine(n));
    }
    catch (AggregateException ex)
    {
        foreach (var e in ex.InnerExceptions)
        {
            Console.WriteLine(e.Message);
        }
    }

    int Square(int n) => n * n;
    ```
   *Ссылка на положения*: [1, 2]

2. **Прерывание параллельной операции с использованием `WithCancellation` и `CancellationToken`**:
    ```csharp
    CancellationTokenSource cts = new CancellationTokenSource();
    new Task(() =>
    {
        Thread.Sleep(400);
        cts.Cancel();
    }).Start();

    try
    {
        int[] numbers = new int[] { 1, 2, 3, 4, 5, 6, 7, 8 };

        var squares = from n in numbers.AsParallel().WithCancellation(cts.Token)
                         select Square(n);

        foreach (var n in squares)
            Console.WriteLine(n);
    }
    catch (OperationCanceledException)
    {
        Console.WriteLine("Операция была прервана");
    }
    catch (AggregateException ex)
    {
        if (ex.InnerExceptions != null)
        {
            foreach (Exception e in ex.InnerExceptions)
                Console.WriteLine(e.Message);
        }
    }
    finally
    {
        cts.Dispose();
    }

    int Square(int n)
    {
        var result = n * n;
        Console.WriteLine($"Квадрат числа {n} равен {result}");
        Thread.Sleep(1000); // имитация долгого вычисления
        return result;
    }
    ```
   *Ссылка на положения*: [3, 4]

### Inner links:
[_ LINQ](1.%20Languages/C-sharp/Базы%20данных/LINQ/_%20LINQ.md)

### Outer links:
https://metanit.com/sharp/tutorial/17.3.php